docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,157 @@
+
+## New components
+* Finalizers:
+  * Like a controller, a finalizer is always running.
+  * Watches for update events that meet two conditions:
+    1. the updated objects has the identifier of the finalizer in ObjectMeta.Finalizers;
+    2. ObjectMeta.DeletionInProgress is updated from false to true.
+  * Applies the finalizing logic to the object in the update event.
+  * Removes itself from ObjectMeta.Finalizers after the finalizing logic is completed.
+  * Because it's possible for the finalizing logic to be applied multiple times, the finalizing logic has to be idempotent.
+
+## Changes to existing components
+* API server:
+  * Deletion handler:
+    * If the `ObjectMeta.Finalizers` of the object being deleted is non-empty, then updates the DeletionInProgress field to true.
+    * If the `ObjectMeta.Finalizers` is empty, then deletes the object.
+  * Update handler:
+    * If the update removes the last finalizer, and the DeletionInProgress is true, then deletes the object from the registry.
+
+## 2B. Enforcing orphaning policty with a finalizer
+## API changes
+```
+type ObjectMeta struct {
+	...
+	ParentReferences []ObjectReference
+    Orphaning bool
+}
+```
+**ObjectMeta.ParentReferences**: same as in [design #1](#api-changes).
+
+**ObjectMeta.Orphaning**: same as in [design #1](#api-changes).
+
+## New components
+* Garbage Collector: 
+  * Almost identical to the [Garbage Collector](#new-components) in design #1.
+
+* Orphaning finalizer: 
+  * Could be part of the garbage collector.
+  * Watches for update events as described in [2A](#new-components-1).
+  * If `ObjectMeta.Orphaning` of the updated object is true, removes the object from the `ParentReferences` of its children.
+    * children can be found via the DAG kept by GC, or by relisting the child resource and checking the ParentReferences field.
+  * Removes the finalizer from the `ObjectMeta.Finalizers` of the parent object.
+
+## Expected race conditions
+* *Controller of the parent resrouce creates a child object after the orphaning finalizer has updated the ParentReferences of other children*: This can be solved by implementing Status.ObservedGeneration in all resources, so that before applying the orphaning logic to children, the orphaning finalizer checks Status.ObservedGeneration to ensure the controller of the parent resource has observed the deletion. Note that in design #1, the same [race condition](#expected-race-conditions) is solved by introducing the tombstone.
+
+# Comparison between design #1 and #2
+* Advantages of "Tombstone + GC" design
+  * Faster compared to using finalizers. The original object can be deleted once the tombstone is created, rather than waiting for the finalizers to update all children’s ObejectMeta.ParentReferences.
+  * Doesn't require implementing ObservedGeneration in all resources.
```

Comment:
Also, I think it's questionable how much we care.

If the controller observes that the object under control is being deleted, then it will stop creating children with parent references.

If the controller doesn't observe that the object is being deleted, then it might create new children with parent references, which would then get deleted quickly after their parent is deleted.

Pretty much the same outcome.